azurerm/resource_arm_key_vault_policy.go

```diff
@@ -0,0 +1,221 @@
+				Required:     true,
+				ValidateFunc: validateUUID,
+				ForceNew:     true,
+			},
+			"application_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateUUID,
+				ForceNew:     true,
+			},
+			"certificate_permissions": certificatePermissionsSchema(),
+			"key_permissions":         keyPermissionsSchema(),
+			"secret_permissions":      secretPermissionsSchema(),
+		},
+	}
+}
+
+func createKeyVaultAccessPolicy(d *schema.ResourceData) *keyvault.AccessPolicyEntry {
+
+	certificatePermissionsRaw := d.Get("certificate_permissions").([]interface{})
+	certificatePermissions := []keyvault.CertificatePermissions{}
+	for _, permission := range certificatePermissionsRaw {
+		certificatePermissions = append(certificatePermissions, keyvault.CertificatePermissions(permission.(string)))
+	}
+
+	keyPermissionsRaw := d.Get("key_permissions").([]interface{})
+	keyPermissions := []keyvault.KeyPermissions{}
+	for _, permission := range keyPermissionsRaw {
+		keyPermissions = append(keyPermissions, keyvault.KeyPermissions(permission.(string)))
+	}
+
+	secretPermissionsRaw := d.Get("secret_permissions").([]interface{})
+	secretPermissions := []keyvault.SecretPermissions{}
+	for _, permission := range secretPermissionsRaw {
+		secretPermissions = append(secretPermissions, keyvault.SecretPermissions(permission.(string)))
+	}
+
+	policy := keyvault.AccessPolicyEntry{
+		Permissions: &keyvault.Permissions{
+			Certificates: &certificatePermissions,
+			Keys:         &keyPermissions,
+			Secrets:      &secretPermissions,
+		},
+	}
+
+	tenantUUID := uuid.FromStringOrNil(d.Get("tenant_id").(string))
+	policy.TenantID = &tenantUUID
+	objectUUID := d.Get("object_id").(string)
+	policy.ObjectID = &objectUUID
+
+	if v := d.Get("application_id"); v != "" {
+		applicationUUID := uuid.FromStringOrNil(v.(string))
+		policy.ApplicationID = &applicationUUID
+	}
+
+	return &policy
+
+}
+
+func resourceArmKeyVaultPolicyCreateOrDelete(d *schema.ResourceData, meta interface{}, action keyvault.AccessPolicyUpdateKind) error {
+	client := meta.(*ArmClient).keyVaultClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for Azure ARM Policy: %s.", action)
+
+	name := d.Get("vault_name").(string)
+	//location := azureRMNormalizeLocation(d.Get("vault_location").(string))
+	resGroup := d.Get("vault_resource_group").(string)
+
+	parameters := keyvault.VaultAccessPolicyParameters{
+		Name: &name,
+		Properties: &keyvault.VaultAccessPolicyProperties{
+			AccessPolicies: &[]keyvault.AccessPolicyEntry{*createKeyVaultAccessPolicy(d)},
+		},
+	}
+
+	_, err := client.UpdateAccessPolicy(ctx, resGroup, name, action, parameters)
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(ctx, resGroup, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("cannot read KeyVault %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmKeyVaultRead(d, meta)
```

Comment:
Typically we don't read terraform resources after it is deleted, and we don't set its ID either.

Reply:
Added a check that we aren't deleting it.